pyro/infer/svgd.py

```diff
@@ -81,51 +77,47 @@ def _bandwidth(self, norm_sq):
         index = torch.arange(num_particles)
         norm_sq = norm_sq[index > index.unsqueeze(-1), ...]
         median = norm_sq.median(dim=0)[0]
+        if self.bandwidth_factor is not None:
+            median = self.bandwidth_factor * median
+        assert median.shape == norm_sq.shape[-1:]
         return median / math.log(num_particles + 1)
 
-    def _log_kernel_and_grad(self, param, bandwidth_factor=None):
-        """
-        Compute the component kernels and (parts of) their gradients.
-
-        :param float bandwidth_factor: optional factor by which to scale the bandwidth
-        """
-        num_particles = param.size(0)
-        delta_x = param.unsqueeze(0) - param.unsqueeze(1)
-        norm_sq = delta_x.reshape(num_particles, num_particles, -1).pow(2.0)
-        h = self._bandwidth(norm_sq)
-        if bandwidth_factor:
-            h *= bandwidth_factor
-        log_kernel = -(norm_sq / h).sum(-1)
-        grad_term = (-2.0) * delta_x / h.reshape(param.shape[1:])
+    @torch.no_grad()
+    def log_kernel_and_grad(self, particles):
+        delta_x = particles.unsqueeze(0) - particles.unsqueeze(1)  # N N D
+        assert delta_x.dim() == 3
+        norm_sq = delta_x.pow(2.0)  # N N D
+        h = self._bandwidth(norm_sq)  # D
+        log_kernel = -(norm_sq / h)  # N N D
+        grad_term = 2.0 * delta_x / h  # N N D
+        assert log_kernel.shape == grad_term.shape
         return log_kernel, grad_term
 
-    def kernel_and_grads(self, params, bandwidth_factor=None):
-        """
-        Compute the full kernel and its gradient
+    @property
+    def bandwidth_factor(self):
+        return self._bandwidth_factor
 
-        :param dict params: dictionary of (name, parameter) pairs
-        :param float bandwidth_factor: optional factor by which to scale the bandwidth
+    @bandwidth_factor.setter
+    def bandwidth_factor(self, bandwidth_factor):
         """
-        log_kernels, grads = {}, {}
-        for name, param in params.items():
-            log_kernel, grad = self._log_kernel_and_grad(param, bandwidth_factor=bandwidth_factor)
-            log_kernels[name] = log_kernel
-            grads[name] = grad
-        kernel = torch.exp(sum(log_kernels.values()))
-        grads = {name: torch.einsum("ji,ji...->i...", kernel, grad) for name, grad in grads.items()}
-        return kernel, grads
+        :param float bandwidth_factor: Optional factor by which to scale the bandwidth
```

Comment:
I'm not sure this shows up in docs. If not, you could add the following the class documentation:
```
:ivar float bandwidth_factor: Optional factor ...
```

Reply:
thanks done